### PR TITLE
ompi/errhandler: fix comm errhandler issue

### LINKED ISSUE #8341: ompi - errhandler fortran test fails
### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
master

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
git clone

### If you are building/installing from a git clone, please copy-n-paste the output from `git submodule status`.
```
[abodevar@f8n02 ompi]$ git submodule status
 cf5b492ce474bb7032f94c633e5f068293e69c2c 3rd-party/openpmix (v1.1.3-2641-gcf5b492c)
 080c116a85667a413097fe987634d3c733b3400f 3rd-party/prrte (dev-30767-g080c116a85)

```

### Please describe the system on which you are running

* Operating system/version: RH8.2
* Computer hardware: ppc64le
* Network type: IB

-----------------------------

## Details of the problem

Here is the minimal reproducer:
https://raw.githubusercontent.com/AboorvaDevarajan/mpi-tests/main/errhandler.F

```
[abodevar@f8n02 src]$ mpif77 errhandler.F -o errhandler
[abodevar@f8n02 src]$ mpirun -np 2 ./errhandler
 name 1: MPI_COMM_WORLD                                                 :
 name 2: Tp� �X���h ��S��%:
 name 1: MPI_COMM_WORLD                                                 :
 name 2: Tp� �X?��h @}2��%:
[abodevar@f8n02 src]$ 
```
In the error handler the comm_name contains garbage value.

Looks like for the fortran case, C to Fortran conversion is not done here:
https://github.com/open-mpi/ompi/blob/master/ompi/errhandler/errhandler_invoke.c#L51

Here is a probable fix:
https://github.com/open-mpi/ompi/pull/8342

With the fix:
```
[abodevar@f8n02 src]$ mpirun -np 2 ./errhandler
 name 1: MPI_COMM_WORLD                                                 :
 name 2: MPI_COMM_SELF                                                  :
 name 1: MPI_COMM_WORLD                                                 :
 name 2: MPI_COMM_SELF                                                  :
```

### PR DESCRIPTION
**Pull request by AboorvaDevarajan**
Signed-off-by: Aboorva Devarajan <abodevar@in.ibm.com>

Fixes #8341

### CODE REVIEW

**Comment by jjhursey** [comment]
ok to test

File: ompi/errhandler/errhandler_invoke.c [hunk 1]
```diff
@@ -48,7 +48,22 @@ int ompi_errhandler_invoke(ompi_errhandler_t *errhandler, void *mpi_object,
         if (state >= OMPI_MPI_STATE_INIT_COMPLETED &&
             state < OMPI_MPI_STATE_FINALIZE_PAST_COMM_SELF_DESTRUCT) {
             comm = (ompi_mpi_compat_mpi3)? &ompi_mpi_comm_world.comm: &ompi_mpi_comm_self.comm;
-            comm->error_handler->eh_comm_fn(&comm, &err_code, message, NULL);
+            switch (comm->error_handler->eh_lang) {
+               case OMPI_ERRHANDLER_LANG_C:
+                 comm->error_handler->eh_comm_fn(&comm, &err_code, message, NULL);
+                 break;
+
+               case OMPI_ERRHANDLER_LANG_CXX:
```
Comment (jsquyres) [inline ompi/errhandler/errhandler_invoke.c:56]: Open MPI doesn't have the C++ bindings any more -- should we remove all references to `OMPI_ERRHANDLER_LANG_CXX`?

Comment (bosilca) [reply ompi/errhandler/errhandler_invoke.c:56]: If we decide to keep the internal C++ support we should at least pass the C++ error handler and not the C one as the last parameter of the callback.

Comment (jsquyres) [reply ompi/errhandler/errhandler_invoke.c:56]: Good point.

I think we should probably just ditch all the C++ glue here -- `OMPI_ERRHANDLER_LANG_CXX`, the `eh_cxx_dispatch_fn`, ... etc.  It's all dead code nowadays.

**Comment by AboorvaDevarajan** [comment]
@jsquyres Jeff, I have removed the stale CXX bindings.

**Comment by jsquyres** [comment]
I just added the following trivial addition to your commit:

```patch
diff --git a/ompi/errhandler/errhandler.h b/ompi/errhandler/errhandler.h
index 4d6cef983f..b00296096a 100644
--- a/ompi/errhandler/errhandler.h
+++ b/ompi/errhandler/errhandler.h
@@ -88,12 +88,6 @@ enum ompi_errhandler_type_t {
 typedef enum ompi_errhandler_type_t ompi_errhandler_type_t;
 
 
-/*
- * Need to forward declare this for use in ompi_errhandle_cxx_dispatch_fn_t.
- */
-struct ompi_errhandler_t;
-
-
 /**
  * Back-end type for MPI_Errorhandler.
  */
```

**Review by jsquyres — APPROVED** [review]

**Comment by jsquyres** [comment]
Sorry -- one last trivial update: I re-ordered the 2 commits: put the "remove C++ stuff" first so that the 2nd commit (that doesn't include C++ stuff) makes more sense in that logical order.

**Comment by jsquyres** [comment]
@AboorvaDevarajan Do you want to cherry pick the 2nd commit to v4.0.x and v4.1.x?  Keep in mind that these 2 branches still have the C++ bindings, so you'll need to re-add the C++ clause you originally had in this PR (sorry about that!).

**Comment by AboorvaDevarajan** [comment]
Looks like only `master` branch has this issue.

**Comment by AboorvaDevarajan** [comment]
Thanks Jeff, Looks like only `master` branch has this issue.

**Comment by jsquyres** [comment]
@AboorvaDevarajan Awesome; thanks for checking.

**Pull request merged**